### PR TITLE
RDM-1240 - search cases by metadata

### PR DESCRIPTION
**Pull request by gbenadikar**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-1240

### Change description ###
1. Definition import process changed to add metadata fields to the parser context to support defining metadata fields in excel tabs.
2. Introduced validations not to allow metadata fields in case field tab and should be read-only on case event fields tab
3. When retrieving case type information, the metadata fields will be added to the case type - case fields with fixed list values if a field is a fixed list
Depends on
https://github.com/hmcts/ccd-definition-store-api/pull/77

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContext.java [hunk 1]
```diff
@@ -3,19 +3,28 @@
...
+import java.util.function.Function;
+
+import static java.util.stream.Collectors.toMap;
 
 /**
  * Accumulate everything that has been parsed so far and which is required for a subsequent parse stage. This is not meant
  * to expose the result of the parse itself, but instead to facilitate the mapping of dependent elements parsed at different
  * stages. As such, it should focus on the ease of consumption by parser.
- *
+ * <p>
```
Comment (vlaurin) [inline excel-importer/src/main/java/uk/gov/hmcts/ccd/definition/store/excel/parser/ParseContext.java]: ?

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/metadata/StateMetadataFixedListItemFactory.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+public class StateMetadataFixedListItemFactory implements MetadataFixedListItemFactory {
+
+    public static final String QUALIFIER = "StateMetadataFixedListItemFactory";
+
+    @Override
+    public List<FixedListItem> createFixedListItems(CaseType caseType) {
+        return caseType.getStates()
+            .stream()
+            .map(state -> new FixedListItem(state.getId(), state.getName()))
+            .collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (gbenadikar) [reply domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/metadata/StateMetadataFixedListItemFactory.java]: To make entries in field_type_list_item for state when importing, there were a lot of code changes to the definition import process. Changes to ListFieldTypeParser to read State tab and create entries (case type specific), FieldTypeParser, CaseFieldParser etc. and was getting a little messy. 

Wanted to keep the changes isolated so thought doing it on retrieval and enriching the list values when adding metadata fields to case fields. As we are caching the case type definition call, will it not be ok to keep the fixed list item creation here.

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/metadata/MetadataFieldServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        // Get metadata fields and convert entity to dto
+        List<CaseField> metadataFields = caseFieldRepository.findByDataFieldType(METADATA)
+            .stream()
+            .map(dtoMapper::map)
+            .collect(toList());
+
+        // enrich metadata fixed lists with values
+        metadataFields.stream()
+            .filter(field -> BASE_FIXED_LIST.equals(field.getFieldType().getType()))
+            .forEach(field -> setFixedListItems(field, caseType));
```
[the remark this answers is not in the document]
Comment (gbenadikar) [reply domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/metadata/MetadataFieldServiceImpl.java]: I think same explanation as the below reasoning for creating fixed list item values for metadata at runtime. We can discuss this further on Monday.

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/metadata/MetadataField.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    LAST_MODIFIED,
+    SECURITY_CLASSIFICATION;
+
+    public static MetadataField fromString(String fieldName) {
+        for (MetadataField metadataField : values()) {
+            if (metadataField.name().equalsIgnoreCase(fieldName)) {
+                return metadataField;
+            }
+        }
+        return null;
```
[the remark this answers is not in the document]
Comment (gbenadikar) [reply domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/service/metadata/MetadataField.java]: I am ignoring the null value safely with ofNullable. Just that if we add a metadata to DB directly without adding the field to the enum, throwing exception will break the case type retrieval call.

**Pull request merged**